protocols/gossipsub/src/config.rs

```diff
@@ -69,8 +70,8 @@ pub struct GossipsubConfig {
     duplicate_cache_time: Duration,
     validate_messages: bool,
     validation_mode: ValidationMode,
-    message_id_fn: fn(&GossipsubMessage) -> MessageId,
-    fast_message_id_fn: Option<fn(&RawGossipsubMessage) -> FastMessageId>,
+    message_id_fn: Arc<dyn Fn(&GossipsubMessage) -> MessageId + Send + Sync + 'static>,
```

Comment:
Could you explain why an `Arc` instead of a `Box<dyn Fn ...>` is needed here?

Reply:
Hi @mxinden 
From the derive above, the `GossipsubConfig` needs to be `Clone`. So if I used a `Box`, I would need to add an extra `Clone` trait bound inside the angle brackets. Because `Fn(&GossipsubMessage) -> MessageId` is already a non-auto trait, this requires us to create a `NewTrait` implementing `Fn<(&'r RawGossipsubMessage,)> + Clone` and then make this a `Box<NewTrait>`.

All of that makes the code less readable, and I'm not actually sure if any closure can be made `Clone` in the first place. But if you take a step back and consider the `Arc<T>` for a moment, the only time you need to pay the synchronization overhead is when the reference count changes (during `Arc<T>::new()`, `Arc<T>::clone()`, and `Arc<T>::drop()`). The [atomic counters aren't touched at all during Deref](https://doc.rust-lang.org/src/alloc/sync.rs.html#1299-1306).

So it should produce no performance hit compared to a `Box` and is actually probably faster than cloning the whole closure each time the `GossipsubConfig` is cloned.